net/portmapper/upnp.go

```diff
@@ -421,38 +423,133 @@ func (c *Client) getUPnPPortMapping(
-			c.logf("getUPnPRootDevice: loc=%q err=%v", loc, err)
+
+	// Wrapper for a uPnPDiscoResponse with an optional existing root
+	// device + URL (if we've got a previous cached mapping).
+	type step struct {
+		rootDev *goupnp.RootDevice // if nil, use 'meta'
+		loc     *url.URL           // non-nil if rootDev is non-nil
+		meta    uPnPDiscoResponse
+	}
+	var steps []step
+
+	// Now, if we have an existing mapping, swap that mapping's entry to
+	// the first entry in our "metas" list so we try it first.
+	haveOldMapping := ok && oldMapping != nil
+	if haveOldMapping && oldMapping.rootDev != nil {
+		steps = append(steps, step{rootDev: oldMapping.rootDev, loc: oldMapping.loc})
+	}
+	// Note: this includes the meta for a previously-cached mapping, in
+	// case the rootDev changes.
+	for _, meta := range metas {
+		steps = append(steps, step{meta: meta})
+	}
+
+	// Now, iterate through every meta that we have trying to get an
+	// external IP address. If we succeed, we'll return; if we fail, we
+	// continue this loop.
+	var errs []error
+	for _, step := range steps {
+		var (
+			rootDev *goupnp.RootDevice
+			loc     *url.URL
+			err     error
+		)
+		if step.rootDev != nil {
+			rootDev = step.rootDev
+			loc = step.loc
+		} else {
+			rootDev, loc, err = getUPnPRootDevice(ctx, c.logf, c.debug, gw, step.meta)
+			c.vlogf("getUPnPRootDevice: loc=%q err=%v", loc, err)
+			if err != nil {
+				errs = append(errs, err)
+				continue
+			}
+		}
+		if rootDev == nil {
+			continue
 		}
+
+		// This actually performs the port mapping operation using this
+		// root device.
+		//
+		// TODO(andrew-d): this can successfully perform a portmap and
+		// return an externalAddrPort that refers to a non-public IP
+		// address if the first selected RootDevice is a device that is
+		// connected to another internal network. This is still better
+		// than randomly flapping between multiple devices, but we
+		// should probably split this up further to try the best
+		// service (one with an external IP) first, instead of
+		// iterating by device.
+		//
+		// This is probably sufficiently unlikely that I'm leaving that
+		// as a follow-up task if it's necessary.
+		externalAddrPort, client, err := c.tryUPnPPortmapWithDevice(ctx, internal, prevPort, rootDev, loc)
 		if err != nil {
-			return netip.AddrPort{}, false
+			errs = append(errs, err)
+			continue
 		}
-	}
-	if rootDev == nil {
-		return netip.AddrPort{}, false
-	}
 
-	// Now that we have a root device, select the best mapping service from
-	// it. This makes network requests, and can vary from mapping to
-	// mapping if the upstream device's connection status changes.
+		// If we get here, we're successful; we can cache this mapping,
+		// update our local port, and then return.
+		//
+		// NOTE: this time might not technically be accurate if we created a
+		// permanent lease above, but we should still re-check the presence of
+		// the lease on a regular basis so we use it anyway.
```

Comment:
Do we try to make a permanent lease? I think it would be better to always try to make one that expires.

Reply:
We always try a temporary lease first, but some UPnP implementations don't support that. In those cases, we fall back to a permanent lease, though we still explicitly remove that lease when we're not using it.